
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
        <link rel="next" href="reference/">
      
      
        
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.7.2">
    
    
      
        <title>Home - baktfold</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.484c7ddc.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#baktfold" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="." title="baktfold" class="md-header__button md-logo" aria-label="baktfold" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            baktfold
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/gbouras13/baktfold" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="." title="baktfold" class="md-nav__button md-logo" aria-label="baktfold" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    baktfold
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/gbouras13/baktfold" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
      <a href="." class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            
  
    Reference
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Index
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/bakta/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Bakta
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/databases/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Databases
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/features/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Features
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/io/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Io
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/prokka_gbk_to_json/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Prokka gbk to json
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/results/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Results
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/subcommands/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Subcommands
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="reference/utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Utils
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/gbouras13/baktfold/edit/master/docs/index.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


<p><a href="https://colab.research.google.com/github/gbouras13/baktfold/blob/main/run_baktfold.ipynb"><img alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg" /></a></p>
<p><a href="https://anaconda.org/bioconda/baktfold"><img alt="Anaconda-Server Badge" src="https://anaconda.org/bioconda/baktfold/badges/version.svg" /></a>
<a href="https://img.shields.io/conda/dn/bioconda/baktfold"><img alt="Bioconda Downloads" src="https://img.shields.io/conda/dn/bioconda/baktfold" /></a>
<a href="https://badge.fury.io/py/baktfold"><img alt="PyPI version" src="https://badge.fury.io/py/baktfold.svg" /></a>
<a href="https://pepy.tech/project/baktfold"><img alt="Downloads" src="https://static.pepy.tech/badge/baktfold" /></a></p>
<h1 id="baktfold">baktfold</h1>
<p>Rapid &amp; standardized annotation of bacterial genomes, MAGs &amp; plasmids using protein structural information</p>
<p><code>baktfold</code> is a sensitive annotation tool for bacterial genomes, MAGs &amp; plasmids genomes using protein structural homology. </p>
<p><code>baktfold</code> is very similar to <a href="https://github.com/gbouras13/phold">phold</a> but goes beyond phages to bacterial annotation. <code>baktfold</code> takes all "hypothetical proteins" from <a href="https://github.com/oschwengers/bakta">Bakta's</a> output and uses the <a href="https://github.com/mheinzinger/ProstT5">ProstT5</a> protein language model to rapidly translate protein amino acid sequences to the 3Di token alphabet used by <a href="https://github.com/steineggerlab/foldseek">Foldseek</a>. Foldseek is then used to search these against a series of databases (SwissProt, AlphaFold Database non-singleton clusters, PDB and CATH).</p>
<p>Additionally, instead of using ProstT5, you can specify protein structures that you have pre-computed for your hypothetical proteins.</p>
<p>You can also specify custom databases to search against using <code>--custom-db</code></p>
<p><strong>Baktfold is currently under active development. We would welcome any and all feedback (especially bugs) via Issues</strong></p>
<h1 id="google-colab-notebook">Google Colab Notebook</h1>
<p>If you don't want to install <code>baktfold</code> locally, you can run it without any code using the <a href="https://colab.research.google.com/github/gbouras13/baktfold/blob/main/run_baktfold.ipynb">Google Colab notebook</a></p>
<h1 id="table-of-contents">Table of Contents</h1>
<ul>
<li><a href="#baktfold">baktfold</a></li>
<li><a href="#google-colab-notebook">Google Colab Notebook</a></li>
<li><a href="#table-of-contents">Table of Contents</a></li>
<li><a href="#install">Install</a><ul>
<li><a href="#conda-recommended">Conda (recommended)</a></li>
<li><a href="#pip">Pip</a></li>
<li><a href="#source">Source</a></li>
<li><a href="#database-installation">Database Installation</a></li>
</ul>
</li>
<li><a href="#example">Example</a></li>
<li><a href="#usage">Usage</a></li>
<li><a href="#output">Output</a><ul>
<li><a href="#conceptual-terms">Conceptual terms</a></li>
</ul>
</li>
<li><a href="#citations">Citations</a></li>
</ul>
<h2 id="install">Install</h2>
<h3 id="conda-recommended">Conda (recommended)</h3>
<ul>
<li>The best way to install <code>baktfold</code> is using conda, as this will install Foldseek (the only non-Python dependency) along with the Python dependencies</li>
<li>We would highly recommend installing conda via <a href="https://github.com/conda-forge/miniforge">miniforge</a>.</li>
<li>To install baktfold:</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>baktfoldENV<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>baktfold<span class="w"> </span>
</span></code></pre></div>
<ul>
<li>To utilise phold with GPU, a GPU compatible version of pytorch must be installed. By default conda will usually install a CPU-only version.</li>
<li>If you have an NVIDIA GPU, please try:</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>baktfoldENV<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>baktfold<span class="w"> </span><span class="nv">pytorch</span><span class="o">=</span>*<span class="o">=</span>cuda*
</span></code></pre></div>
<ul>
<li>If you have a Mac with M-series Apple Silicon, you may need to install a particular version of Pytorch to utilise GPU-acceleration</li>
<li>The same is true if you use other non-NVIDIA e.g. AMD GPUs</li>
<li>See <a href="https://phold.readthedocs.io/en/latest/install/">this link</a> for some more detail and further links</li>
</ul>
<h3 id="pip">Pip</h3>
<ul>
<li>You can also install baktfold using pip.</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>pip<span class="w"> </span>install<span class="w"> </span>baktfold
</span></code></pre></div>
<ul>
<li>You will need to have Foldseek (ideally v10.941cd33) installed and available in the $PATH.</li>
</ul>
<h3 id="source">Source</h3>
<ul>
<li>You can install the latest version of baktfold with potentially untested and unreleased changes into a conda environment using conda as follows:</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>conda<span class="w"> </span>create<span class="w"> </span>-n<span class="w"> </span>baktfoldENV<span class="w"> </span>foldseek
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>conda<span class="w"> </span>activate<span class="w"> </span>baktfoldENV
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/gbouras13/baktfold.git
</span><span id="__span-3-4"><a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="nb">cd</span><span class="w"> </span>baktfold
</span><span id="__span-3-5"><a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a>pip<span class="w"> </span>install<span class="w"> </span>.
</span><span id="__span-3-6"><a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a>baktfold<span class="w"> </span>--help
</span></code></pre></div>
<h3 id="database-installation">Database Installation</h3>
<ul>
<li>To download and install baktfold's databases (use as many threads with <code>-t</code> as you can to speed up downloading)</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>baktfold<span class="w"> </span>install<span class="w"> </span>-d<span class="w"> </span>baktfold_db<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span>
</span></code></pre></div>
<ul>
<li>If you have an NVIDIA GPU, you will need to format the database to allow it to use Foldseek-GPU with <code>--foldseek-gpu</code><ul>
<li>Note: you can do this after downloading the database with the above command (it won't redownload the database, only do the relevant Foldseek database padding)</li>
</ul>
</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>baktfold<span class="w"> </span>install<span class="w"> </span>-d<span class="w"> </span>baktfold_db<span class="w"> </span>--foldseek-gpu
</span></code></pre></div>
<h2 id="example">Example</h2>
<ul>
<li>You will first need to run <a href="https://github.com/oschwengers/bakta">bakta</a> and use the resulting <code>.json</code> file as input for <code>baktfold</code><ul>
<li>We will add other input formats eventually, put we would always recommend running Bakta first, as it is awesome and comprehensive</li>
</ul>
</li>
<li>To use <code>baktfold run</code> using a dummy test example</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="c1"># with nvidia gpu </span>
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>baktfold<span class="w"> </span>run<span class="w"> </span>-i<span class="w"> </span>tests/test_data/assembly_bakta_output/assembly.json<span class="w">  </span>-o<span class="w"> </span>baktfold_output<span class="w"> </span>-f<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-d<span class="w"> </span>baktfold_db<span class="w">   </span>--foldseek-gpu
</span><span id="__span-6-3"><a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="c1"># without nvidia gpu available</span>
</span><span id="__span-6-4"><a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a>baktfold<span class="w"> </span>run<span class="w"> </span>-i<span class="w"> </span>tests/test_data/assembly_bakta_output/assembly.json<span class="w">  </span>-o<span class="w"> </span>baktfold_output<span class="w"> </span>-f<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-d<span class="w"> </span>baktfold_db<span class="w">   </span>
</span></code></pre></div>
<ul>
<li>To use <code>baktfold proteins</code> using a dummy test example protein <code>.faa</code> file</li>
<li>Note that this can be any <code>.faa</code> (It does not have to be the output of Bakta)</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="c1"># with nvidia gpu </span>
</span><span id="__span-7-2"><a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>baktfold<span class="w"> </span>proteins<span class="w"> </span>-i<span class="w"> </span>tests/test_data/assembly.hypotheticals.faa<span class="w">  </span>-o<span class="w"> </span>baktfold_proteins_output<span class="w"> </span>-f<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-d<span class="w"> </span>baktfold_db<span class="w">   </span>--foldseek-gpu
</span><span id="__span-7-3"><a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="c1"># without nvidia gpu available</span>
</span><span id="__span-7-4"><a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a>baktfold<span class="w"> </span>proteins<span class="w"> </span>-i<span class="w"> </span>tests/test_data/assembly.hypotheticals.faa<span class="w">  </span>-o<span class="w"> </span>baktfold_proteins_output<span class="w"> </span>-f<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>-d<span class="w"> </span>baktfold_db<span class="w">   </span>
</span></code></pre></div>
<h2 id="usage">Usage</h2>
<ul>
<li>The two most useful commands are <code>baktfold run</code> and <code>baktfold proteins</code></li>
<li><code>baktfold run</code> accepts a <strong>Bakta json file</strong> as its input, and by default, it will annotate all hypothetical CDS and return a variety of Bakta-like compliant output formats. All other annotations will be inherited from the Bakta output</li>
<li><code>baktfold proteins</code> accepts a protein FASTA <code>.faa</code> format file as input. It will annotate all protein sequences and return a variety of <code>bakta_proteins</code>-like output formats</li>
<li>
<p><code>baktfold predict</code> and <code>baktfold compare</code> split <code>baktfold run</code> into the ProstT5 and Foldseek modules, while <code>baktfold proteins-predict</code> and <code>baktfold proteins-compare</code> do the same for <code>baktfold proteins</code> (useful if you have non-NVIDIA GPUs)</p>
</li>
<li>
<p>It is recommend you run baktfold with a GPU if you can.</p>
</li>
<li>If you do not have a GPU, baktfold will still run, but the ProstT5 step will be fairly slow.</li>
<li>If you have a NVIDIA GPU, you can also use the <code>--foldseek-gpu</code> parameter to accelerate Foldseek further</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>Usage:<span class="w"> </span>baktfold<span class="w"> </span><span class="o">[</span>OPTIONS<span class="o">]</span><span class="w"> </span>COMMAND<span class="w"> </span><span class="o">[</span>ARGS<span class="o">]</span>...
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>Options:
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="w">  </span>-h,<span class="w"> </span>--help<span class="w">     </span>Show<span class="w"> </span>this<span class="w"> </span>message<span class="w"> </span>and<span class="w"> </span>exit.
</span><span id="__span-8-5"><a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a><span class="w">  </span>-V,<span class="w"> </span>--version<span class="w">  </span>Show<span class="w"> </span>the<span class="w"> </span>version<span class="w"> </span>and<span class="w"> </span>exit.
</span><span id="__span-8-6"><a id="__codelineno-8-6" name="__codelineno-8-6" href="#__codelineno-8-6"></a>
</span><span id="__span-8-7"><a id="__codelineno-8-7" name="__codelineno-8-7" href="#__codelineno-8-7"></a>Commands:
</span><span id="__span-8-8"><a id="__codelineno-8-8" name="__codelineno-8-8" href="#__codelineno-8-8"></a><span class="w">  </span>citation<span class="w">          </span>Print<span class="w"> </span>the<span class="w"> </span>citation<span class="o">(</span>s<span class="o">)</span><span class="w"> </span><span class="k">for</span><span class="w"> </span>this<span class="w"> </span>tool
</span><span id="__span-8-9"><a id="__codelineno-8-9" name="__codelineno-8-9" href="#__codelineno-8-9"></a><span class="w">  </span>compare<span class="w">           </span>Runs<span class="w"> </span>Foldseek<span class="w"> </span>vs<span class="w"> </span>baktfold<span class="w"> </span>db
</span><span id="__span-8-10"><a id="__codelineno-8-10" name="__codelineno-8-10" href="#__codelineno-8-10"></a><span class="w">  </span>install<span class="w">           </span>Installs<span class="w"> </span>ProstT5<span class="w"> </span>model<span class="w"> </span>and<span class="w"> </span>baktfold<span class="w"> </span>database
</span><span id="__span-8-11"><a id="__codelineno-8-11" name="__codelineno-8-11" href="#__codelineno-8-11"></a><span class="w">  </span>predict<span class="w">           </span>Uses<span class="w"> </span>ProstT5<span class="w"> </span>to<span class="w"> </span>predict<span class="w"> </span>3Di<span class="w"> </span>tokens<span class="w"> </span>-<span class="w"> </span>GPU<span class="w"> </span>recommended
</span><span id="__span-8-12"><a id="__codelineno-8-12" name="__codelineno-8-12" href="#__codelineno-8-12"></a><span class="w">  </span>proteins<span class="w">          </span>baktfold<span class="w"> </span>protein-predict<span class="w"> </span><span class="k">then</span><span class="w"> </span>comapare<span class="w"> </span>all<span class="w"> </span><span class="k">in</span><span class="w"> </span>one<span class="w"> </span>-<span class="w"> </span>GPU...
</span><span id="__span-8-13"><a id="__codelineno-8-13" name="__codelineno-8-13" href="#__codelineno-8-13"></a><span class="w">  </span>proteins-compare<span class="w">  </span>Runs<span class="w"> </span>Foldseek<span class="w"> </span>vs<span class="w"> </span>baktfold<span class="w"> </span>db<span class="w"> </span>on<span class="w"> </span>proteins<span class="w"> </span>input
</span><span id="__span-8-14"><a id="__codelineno-8-14" name="__codelineno-8-14" href="#__codelineno-8-14"></a><span class="w">  </span>proteins-predict<span class="w">  </span>Runs<span class="w"> </span>ProstT5<span class="w"> </span>on<span class="w"> </span>a<span class="w"> </span>multiFASTA<span class="w"> </span>input<span class="w"> </span>-<span class="w"> </span>GPU<span class="w"> </span>recommended
</span><span id="__span-8-15"><a id="__codelineno-8-15" name="__codelineno-8-15" href="#__codelineno-8-15"></a><span class="w">  </span>run<span class="w">               </span>baktfold<span class="w"> </span>predict<span class="w"> </span><span class="k">then</span><span class="w"> </span>comapare<span class="w"> </span>all<span class="w"> </span><span class="k">in</span><span class="w"> </span>one<span class="w"> </span>-<span class="w"> </span>GPU...
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>Usage:<span class="w"> </span>baktfold<span class="w"> </span>run<span class="w"> </span><span class="o">[</span>OPTIONS<span class="o">]</span>
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
</span><span id="__span-9-3"><a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="w">  </span>baktfold<span class="w"> </span>predict<span class="w"> </span><span class="k">then</span><span class="w"> </span>comapare<span class="w"> </span>all<span class="w"> </span><span class="k">in</span><span class="w"> </span>one<span class="w"> </span>-<span class="w"> </span>GPU<span class="w"> </span>recommended
</span><span id="__span-9-4"><a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>
</span><span id="__span-9-5"><a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a>Options:
</span><span id="__span-9-6"><a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="w">  </span>-h,<span class="w"> </span>--help<span class="w">                     </span>Show<span class="w"> </span>this<span class="w"> </span>message<span class="w"> </span>and<span class="w"> </span>exit.
</span><span id="__span-9-7"><a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a><span class="w">  </span>-V,<span class="w"> </span>--version<span class="w">                  </span>Show<span class="w"> </span>the<span class="w"> </span>version<span class="w"> </span>and<span class="w"> </span>exit.
</span><span id="__span-9-8"><a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a><span class="w">  </span>-i,<span class="w"> </span>--input<span class="w"> </span>PATH<span class="w">               </span>Path<span class="w"> </span>to<span class="w"> </span>input<span class="w"> </span>file<span class="w"> </span><span class="k">in</span><span class="w"> </span>Bakta<span class="w"> </span>Genbank<span class="w"> </span>format<span class="w"> </span>or
</span><span id="__span-9-9"><a id="__codelineno-9-9" name="__codelineno-9-9" href="#__codelineno-9-9"></a><span class="w">                                 </span>Bakta<span class="w"> </span>JSON<span class="w"> </span>format<span class="w">  </span><span class="o">[</span>required<span class="o">]</span>
</span><span id="__span-9-10"><a id="__codelineno-9-10" name="__codelineno-9-10" href="#__codelineno-9-10"></a><span class="w">  </span>-o,<span class="w"> </span>--output<span class="w"> </span>PATH<span class="w">              </span>Output<span class="w"> </span>directory<span class="w">   </span><span class="o">[</span>default:<span class="w"> </span>output_baktfold<span class="o">]</span>
</span><span id="__span-9-11"><a id="__codelineno-9-11" name="__codelineno-9-11" href="#__codelineno-9-11"></a><span class="w">  </span>-t,<span class="w"> </span>--threads<span class="w"> </span>INTEGER<span class="w">          </span>Number<span class="w"> </span>of<span class="w"> </span>threads<span class="w">  </span><span class="o">[</span>default:<span class="w"> </span><span class="m">1</span><span class="o">]</span>
</span><span id="__span-9-12"><a id="__codelineno-9-12" name="__codelineno-9-12" href="#__codelineno-9-12"></a><span class="w">  </span>-p,<span class="w"> </span>--prefix<span class="w"> </span>TEXT<span class="w">              </span>Prefix<span class="w"> </span><span class="k">for</span><span class="w"> </span>output<span class="w"> </span>files<span class="w">  </span><span class="o">[</span>default:<span class="w"> </span>baktfold<span class="o">]</span>
</span><span id="__span-9-13"><a id="__codelineno-9-13" name="__codelineno-9-13" href="#__codelineno-9-13"></a><span class="w">  </span>-d,<span class="w"> </span>--database<span class="w"> </span>TEXT<span class="w">            </span>Specific<span class="w"> </span>path<span class="w"> </span>to<span class="w"> </span>installed<span class="w"> </span>baktfold<span class="w"> </span>database
</span><span id="__span-9-14"><a id="__codelineno-9-14" name="__codelineno-9-14" href="#__codelineno-9-14"></a><span class="w">  </span>-f,<span class="w"> </span>--force<span class="w">                    </span>Force<span class="w"> </span>overwrites<span class="w"> </span>the<span class="w"> </span>output<span class="w"> </span>directory
</span><span id="__span-9-15"><a id="__codelineno-9-15" name="__codelineno-9-15" href="#__codelineno-9-15"></a><span class="w">  </span>--batch-size<span class="w"> </span>INTEGER<span class="w">           </span>batch<span class="w"> </span>size<span class="w"> </span><span class="k">for</span><span class="w"> </span>ProstT5.<span class="w"> </span><span class="m">1</span><span class="w"> </span>is<span class="w"> </span>usually<span class="w"> </span>fastest.
</span><span id="__span-9-16"><a id="__codelineno-9-16" name="__codelineno-9-16" href="#__codelineno-9-16"></a><span class="w">                                 </span><span class="o">[</span>default:<span class="w"> </span><span class="m">1</span><span class="o">]</span>
</span><span id="__span-9-17"><a id="__codelineno-9-17" name="__codelineno-9-17" href="#__codelineno-9-17"></a><span class="w">  </span>--cpu<span class="w">                          </span>Use<span class="w"> </span>cpus<span class="w"> </span>only.
</span><span id="__span-9-18"><a id="__codelineno-9-18" name="__codelineno-9-18" href="#__codelineno-9-18"></a><span class="w">  </span>--omit-probs<span class="w">                   </span>Do<span class="w"> </span>not<span class="w"> </span>output<span class="w"> </span>per<span class="w"> </span>residue<span class="w"> </span>3Di<span class="w"> </span>probabilities
</span><span id="__span-9-19"><a id="__codelineno-9-19" name="__codelineno-9-19" href="#__codelineno-9-19"></a><span class="w">                                 </span>from<span class="w"> </span>ProstT5.<span class="w"> </span>Mean<span class="w"> </span>per<span class="w"> </span>protein<span class="w"> </span>3Di
</span><span id="__span-9-20"><a id="__codelineno-9-20" name="__codelineno-9-20" href="#__codelineno-9-20"></a><span class="w">                                 </span>probabilities<span class="w"> </span>will<span class="w"> </span>always<span class="w"> </span>be<span class="w"> </span>output.
</span><span id="__span-9-21"><a id="__codelineno-9-21" name="__codelineno-9-21" href="#__codelineno-9-21"></a><span class="w">  </span>--save-per-residue-embeddings<span class="w">  </span>Save<span class="w"> </span>the<span class="w"> </span>ProstT5<span class="w"> </span>embeddings<span class="w"> </span>per<span class="w"> </span>resuide<span class="w"> </span><span class="k">in</span><span class="w"> </span>a
</span><span id="__span-9-22"><a id="__codelineno-9-22" name="__codelineno-9-22" href="#__codelineno-9-22"></a><span class="w">                                 </span>h5<span class="w"> </span>file
</span><span id="__span-9-23"><a id="__codelineno-9-23" name="__codelineno-9-23" href="#__codelineno-9-23"></a><span class="w">  </span>--save-per-protein-embeddings<span class="w">  </span>Save<span class="w"> </span>the<span class="w"> </span>ProstT5<span class="w"> </span>embeddings<span class="w"> </span>as<span class="w"> </span>means<span class="w"> </span>per
</span><span id="__span-9-24"><a id="__codelineno-9-24" name="__codelineno-9-24" href="#__codelineno-9-24"></a><span class="w">                                 </span>protein<span class="w"> </span><span class="k">in</span><span class="w"> </span>a<span class="w"> </span>h5<span class="w"> </span>file
</span><span id="__span-9-25"><a id="__codelineno-9-25" name="__codelineno-9-25" href="#__codelineno-9-25"></a><span class="w">  </span>--mask-threshold<span class="w"> </span>FLOAT<span class="w">         </span>Masks<span class="w"> </span>3Di<span class="w"> </span>residues<span class="w"> </span>below<span class="w"> </span>this<span class="w"> </span>value<span class="w"> </span>of
</span><span id="__span-9-26"><a id="__codelineno-9-26" name="__codelineno-9-26" href="#__codelineno-9-26"></a><span class="w">                                 </span>ProstT5<span class="w"> </span>confidence<span class="w"> </span><span class="k">for</span><span class="w"> </span>Foldseek<span class="w"> </span>searches
</span><span id="__span-9-27"><a id="__codelineno-9-27" name="__codelineno-9-27" href="#__codelineno-9-27"></a><span class="w">                                 </span><span class="o">[</span>default:<span class="w"> </span><span class="m">25</span><span class="o">]</span>
</span><span id="__span-9-28"><a id="__codelineno-9-28" name="__codelineno-9-28" href="#__codelineno-9-28"></a><span class="w">  </span>-e,<span class="w"> </span>--evalue<span class="w"> </span>FLOAT<span class="w">             </span>Evalue<span class="w"> </span>threshold<span class="w"> </span><span class="k">for</span><span class="w"> </span>Foldseek<span class="w">  </span><span class="o">[</span>default:
</span><span id="__span-9-29"><a id="__codelineno-9-29" name="__codelineno-9-29" href="#__codelineno-9-29"></a><span class="w">                                 </span>1e-3<span class="o">]</span>
</span><span id="__span-9-30"><a id="__codelineno-9-30" name="__codelineno-9-30" href="#__codelineno-9-30"></a><span class="w">  </span>-s,<span class="w"> </span>--sensitivity<span class="w"> </span>FLOAT<span class="w">        </span>Sensitivity<span class="w"> </span>parameter<span class="w"> </span><span class="k">for</span><span class="w"> </span>foldseek<span class="w">  </span><span class="o">[</span>default:
</span><span id="__span-9-31"><a id="__codelineno-9-31" name="__codelineno-9-31" href="#__codelineno-9-31"></a><span class="w">                                 </span><span class="m">9</span>.5<span class="o">]</span>
</span><span id="__span-9-32"><a id="__codelineno-9-32" name="__codelineno-9-32" href="#__codelineno-9-32"></a><span class="w">  </span>--keep-tmp-files<span class="w">               </span>Keep<span class="w"> </span>temporary<span class="w"> </span>intermediate<span class="w"> </span>files,
</span><span id="__span-9-33"><a id="__codelineno-9-33" name="__codelineno-9-33" href="#__codelineno-9-33"></a><span class="w">                                 </span>particularly<span class="w"> </span>the<span class="w"> </span>large<span class="w"> </span>foldseek_results.tsv
</span><span id="__span-9-34"><a id="__codelineno-9-34" name="__codelineno-9-34" href="#__codelineno-9-34"></a><span class="w">                                 </span>of<span class="w"> </span>all<span class="w"> </span>Foldseek<span class="w"> </span>hits
</span><span id="__span-9-35"><a id="__codelineno-9-35" name="__codelineno-9-35" href="#__codelineno-9-35"></a><span class="w">  </span>--max-seqs<span class="w"> </span>INTEGER<span class="w">             </span>Maximum<span class="w"> </span>results<span class="w"> </span>per<span class="w"> </span>query<span class="w"> </span>sequence<span class="w"> </span>allowed<span class="w"> </span>to
</span><span id="__span-9-36"><a id="__codelineno-9-36" name="__codelineno-9-36" href="#__codelineno-9-36"></a><span class="w">                                 </span>pass<span class="w"> </span>the<span class="w"> </span>prefilter.<span class="w"> </span>You<span class="w"> </span>may<span class="w"> </span>want<span class="w"> </span>to<span class="w"> </span>reduce
</span><span id="__span-9-37"><a id="__codelineno-9-37" name="__codelineno-9-37" href="#__codelineno-9-37"></a><span class="w">                                 </span>this<span class="w"> </span>to<span class="w"> </span>save<span class="w"> </span>disk<span class="w"> </span>space<span class="w"> </span><span class="k">for</span><span class="w"> </span>enormous<span class="w"> </span>datasets
</span><span id="__span-9-38"><a id="__codelineno-9-38" name="__codelineno-9-38" href="#__codelineno-9-38"></a><span class="w">                                 </span><span class="o">[</span>default:<span class="w"> </span><span class="m">1000</span><span class="o">]</span>
</span><span id="__span-9-39"><a id="__codelineno-9-39" name="__codelineno-9-39" href="#__codelineno-9-39"></a><span class="w">  </span>--ultra-sensitive<span class="w">              </span>Runs<span class="w"> </span>baktfold<span class="w"> </span>with<span class="w"> </span>maximum<span class="w"> </span>sensitivity<span class="w"> </span>by
</span><span id="__span-9-40"><a id="__codelineno-9-40" name="__codelineno-9-40" href="#__codelineno-9-40"></a><span class="w">                                 </span>skipping<span class="w"> </span>Foldseek<span class="w"> </span>prefilter.<span class="w"> </span>Not<span class="w"> </span>recommended
</span><span id="__span-9-41"><a id="__codelineno-9-41" name="__codelineno-9-41" href="#__codelineno-9-41"></a><span class="w">                                 </span><span class="k">for</span><span class="w"> </span>large<span class="w"> </span>datasets.
</span><span id="__span-9-42"><a id="__codelineno-9-42" name="__codelineno-9-42" href="#__codelineno-9-42"></a><span class="w">  </span>--extra-foldseek-params<span class="w"> </span>TEXT<span class="w">   </span>Extra<span class="w"> </span>foldseek<span class="w"> </span>search<span class="w"> </span>params
</span><span id="__span-9-43"><a id="__codelineno-9-43" name="__codelineno-9-43" href="#__codelineno-9-43"></a><span class="w">  </span>--custom-db<span class="w"> </span>TEXT<span class="w">               </span>Path<span class="w"> </span>to<span class="w"> </span>custom<span class="w"> </span>database
</span><span id="__span-9-44"><a id="__codelineno-9-44" name="__codelineno-9-44" href="#__codelineno-9-44"></a><span class="w">  </span>--foldseek-gpu<span class="w">                 </span>Use<span class="w"> </span>this<span class="w"> </span>to<span class="w"> </span><span class="nb">enable</span><span class="w"> </span>compatibility<span class="w"> </span>with
</span><span id="__span-9-45"><a id="__codelineno-9-45" name="__codelineno-9-45" href="#__codelineno-9-45"></a><span class="w">                                 </span>Foldseek-GPU<span class="w"> </span>search<span class="w"> </span>acceleration
</span><span id="__span-9-46"><a id="__codelineno-9-46" name="__codelineno-9-46" href="#__codelineno-9-46"></a><span class="w">  </span>--custom-annotations<span class="w"> </span>PATH<span class="w">      </span>Custom<span class="w"> </span>Foldseek<span class="w"> </span>DB<span class="w"> </span>annotations,<span class="w"> </span><span class="m">2</span><span class="w"> </span>column<span class="w"> </span>tsv.
</span><span id="__span-9-47"><a id="__codelineno-9-47" name="__codelineno-9-47" href="#__codelineno-9-47"></a><span class="w">                                 </span>Column<span class="w"> </span><span class="m">1</span><span class="w"> </span>matches<span class="w"> </span>the<span class="w"> </span>Foldseek<span class="w"> </span>headers,<span class="w"> </span>column
</span><span id="__span-9-48"><a id="__codelineno-9-48" name="__codelineno-9-48" href="#__codelineno-9-48"></a><span class="w">                                 </span><span class="m">2</span><span class="w"> </span>is<span class="w"> </span>the<span class="w"> </span>description.
</span><span id="__span-9-49"><a id="__codelineno-9-49" name="__codelineno-9-49" href="#__codelineno-9-49"></a><span class="w">  </span>-a,<span class="w"> </span>--all-proteins<span class="w">             </span>annotate<span class="w"> </span>all<span class="w"> </span>proteins<span class="w"> </span><span class="o">(</span>not<span class="w"> </span>just
</span><span id="__span-9-50"><a id="__codelineno-9-50" name="__codelineno-9-50" href="#__codelineno-9-50"></a><span class="w">                                 </span>hypotheticals<span class="o">)</span>
</span></code></pre></div>
<h2 id="output">Output</h2>
<ul>
<li>The majority of outputs match <a href="https://github.com/oschwengers/bakta?tab=readme-ov-file#input-and-output">bakta</a>.</li>
<li>Specifically, all the format compliant outputs match bakta's.</li>
<li>The differences are:<ul>
<li><code>&lt;prefix&gt;.inference.tsv</code> is changed compared to bakta. In <code>baktfold</code>, this file gives a quick overview of the different <code>baktfold</code> databases the query protein has hit (if any)</li>
<li>For example:</li>
</ul>
</li>
</ul>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>ID<span class="w">  </span>Length<span class="w">  </span>Product<span class="w"> </span>Swissprot<span class="w">   </span>AFDBClusters<span class="w">    </span>PDB<span class="w"> </span>CATH
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>MEGJMNBEGN_27<span class="w">   </span><span class="m">162</span><span class="w"> </span>HTH-type<span class="w"> </span>quorum-sensing<span class="w"> </span>regulator<span class="w"> </span>RhlR<span class="w">  </span>swissprot_P54292<span class="w">    </span>afdbclusters_A0A9E1VSB0<span class="w"> </span>pdb_5l09<span class="w">    </span>cath_3sztB01
</span><span id="__span-10-3"><a id="__codelineno-10-3" name="__codelineno-10-3" href="#__codelineno-10-3"></a>MEGJMNBEGN_30<span class="w">   </span><span class="m">68</span><span class="w">  </span>hypothetical<span class="w"> </span>protein<span class="w">                </span>
</span><span id="__span-10-4"><a id="__codelineno-10-4" name="__codelineno-10-4" href="#__codelineno-10-4"></a>MEGJMNBEGN_70<span class="w">   </span><span class="m">94</span><span class="w">  </span>hypothetical<span class="w"> </span>protein<span class="w">        </span>afdbclusters_A0A1I3V7E0<span class="w">     </span>
</span></code></pre></div>
<div class="language-text highlight"><pre><span></span><code>* `&lt;prefix&gt;_&lt;database&gt;_tophit.tsv` files give the detailed Foldseek alignment information for each tophit found for each database.
* For example:
</code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>query<span class="w">   </span>target<span class="w">  </span>bitscore<span class="w">    </span>fident<span class="w">  </span>evalue<span class="w">  </span>qStart<span class="w">  </span>qEnd<span class="w">    </span>qLen<span class="w">    </span>qCov<span class="w">    </span>tStart<span class="w">  </span>tEnd<span class="w">    </span>tLen<span class="w">    </span>tCov
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>MEGJMN_070<span class="w">  </span>AF-A0A1I3V7E0-F1-model_v6<span class="w">   </span><span class="m">292</span><span class="w"> </span><span class="m">0</span>.41<span class="w">    </span><span class="m">2</span>.619e-06<span class="w">   </span><span class="m">1</span><span class="w">   </span><span class="m">91</span><span class="w">  </span><span class="m">93</span><span class="w">  </span><span class="m">0</span>.97<span class="w">    </span><span class="m">1</span><span class="w">   </span><span class="m">95</span><span class="w">  </span><span class="m">99</span><span class="w">  </span><span class="m">0</span>.95
</span></code></pre></div>
<div class="language-text highlight"><pre><span></span><code>* The full Foldseek search outputs are not kept by default (only tophits) - you can keep the full Foldseek search tsvs using `--keep-tmp-files`. They will be called `foldseek_results_&lt;database&gt;.tsv`
* `baktfold_3di.fasta` which gives the 3Di tokens for each input CDS
* `baktfold_prostT5_3di_mean_probabilities.csv` and `baktfold_prostT5_3di_all_probabilities.json`, which give some score of the confidence ProstT5 has in its predictions. You can disable this output with `--omit-probs`
* Baktfold does not have plotting functionality like Bakta (yet)
</code></pre></div>
<h3 id="conceptual-terms">Conceptual terms</h3>
<ul>
<li>As Baktfold inherits annotations from Bakta, please see the explanation in bakta for all other concepts <a href="https://github.com/oschwengers/bakta?tab=readme-ov-file#annotation-workflow">here</a></li>
<li>Baktfold adds one conceptual term in addition to Bakta's:<ul>
<li>PSTC: protein structure clusters. These comprise of structure-based annotations to any of Baktfold's databases</li>
</ul>
</li>
</ul>
<h2 id="citations">Citations</h2>
<ul>
<li>
<p>A manuscript describing <code>baktfold</code> is in preparation.</p>
</li>
<li>
<p>Please be sure to cite the following core dependencies - citing all bioinformatics tools that you use helps us, so helps you get better bioinformatics tools:</p>
</li>
<li>
<p>Foldseek - (https://github.com/steineggerlab/foldseek) van Kempen M, Kim S, Tumescheit C, Mirdita M, Lee J, Gilchrist C, Sding J, and Steinegger M. Fast and accurate protein structure search with Foldseek. Nature Biotechnology (2023), <a href="https://www.nature.com/articles/s41587-023-01773-0">doi:10.1038/s41587-023-01773-0 </a></p>
</li>
<li>
<p>ProstT5 - (https://github.com/mheinzinger/ProstT5) Michael Heinzinger, Konstantin Weissenow, Joaquin Gomez Sanchez, Adrian Henkel, Martin Steinegger, Burkhard Rost. ProstT5: Bilingual language model for protein sequence and structure. NAR Genomics and Bioinformatics (2024) <a href="https://doi.org/10.1093/nargab/lqae150">doi:10.1101/2023.07.23.550085</a> </p>
</li>
<li>
<p>Please also consider citing these databases where relevant:</p>
</li>
<li>
<p>AFDB/SwissProt - Mihaly Varadi, Damian Bertoni, Paulyna Magana, Urmila Paramval, Ivanna Pidruchna, Malarvizhi Radhakrishnan, Maxim Tsenkov, Sreenath Nair, Milot Mirdita, Jingi Yeo, Oleg Kovalevskiy, Kathryn Tunyasuvunakool, Agata Laydon, Augustin dek, Hamish Tomlinson, Dhavanthi Hariharan, Josh Abrahamson, Tim Green, John Jumper, Ewan Birney, Martin Steinegger, Demis Hassabis, Sameer Velankar, AlphaFold Protein Structure Database in 2024: providing structure coverage for over 214 million protein sequences, Nucleic Acids Research, Volume 52, Issue D1, 5 January 2024, Pages D368D375, <a href="https://doi.org/10.1093/nar/gkad1011">https://doi.org/10.1093/nar/gkad1011</a></p>
</li>
<li>CATH - Orengo CA, Michie AD, Jones S, Jones DT, Swindells MB, Thornton JM. CATH--a hierarchic classification of protein domain structures. Structure. 1997 Aug 15;5(8):1093-108. doi: 10.1016/s0969-2126(97)00260-8. PMID: 9309224.</li>
<li>PDB - H.M. Berman, J. Westbrook, Z. Feng, G. Gilliland, T.N. Bhat, H. Weissig, I.N. Shindyalov, P.E. Bourne, The Protein Data Bank (2000) Nucleic Acids Research 28: 235-242 <a href="https://doi.org/10.1093/nar/28.1.235">https://doi.org/10.1093/nar/28.1.235</a></li>
</ul>







  
  






                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": ".", "features": ["toc.follow", "content.action.edit"], "search": "assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="assets/javascripts/bundle.79ae519e.min.js"></script>
      
    
  </body>
</html>